File: app/src/main/java/com/udacity/maluleque/meutako/adapters/TransactionsAdapter.java
package com.udacity.maluleque.meutako.adapters;

import android.content.Context;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.core.content.ContextCompat;

import com.udacity.maluleque.meutako.R;
import com.udacity.maluleque.meutako.model.Transaction;
import com.udacity.maluleque.meutako.utils.DateUtils;
import com.udacity.maluleque.meutako.utils.NumberUtils;

import org.zakariya.stickyheaders.SectioningAdapter;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class TransactionsAdapter extends SectioningAdapter {

    private final Context context;
    private List<Transaction> transactions;
    private ArrayList<Section> sections;

    private OnTransactionClickListener onTransactionClickListener;

    public TransactionsAdapter(Context context, List<Transaction> transactions, OnTransactionClickListener onTransactionClickListener) {
        this.transactions = transactions;
        this.sections = new ArrayList<>();
        this.onTransactionClickListener = onTransactionClickListener;
        this.context = context;
        preecheSeccoes();
    }


    @Override
    public SectioningAdapter.ItemViewHolder onCreateItemViewHolder(ViewGroup parent, int itemType) {

        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View v = inflater.inflate(R.layout.transaction_item, parent, false);
        return new TransactionItemViewHolder(v);

    }

    @Override
    public HeaderSectionViewHolder onCreateHeaderViewHolder(ViewGroup parent, int headerType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View v = inflater.inflate(R.layout.header, parent, false);
        return new HeaderSectionViewHolder(v);
    }



    private void preecheSeccoes() {
        List<String> list = new ArrayList<>(new HashSet<>(datesSectionList()));
        Collections.sort(list, new Comparator<String>() {
            DateFormat f = new SimpleDateFormat("MMM yyyy");

            @Override
            public int compare(String o1, String o2) {
                try {
                    return f.parse(o2).compareTo(f.parse(o1));
                } catch (ParseException e) {
                    throw new IllegalArgumentException(e);
                }
            }
        });

        for (int i = 0; i < list.size(); i++) {
            Section section = new Section();
            section.index = i;
            section.header = list.get(i);
            sections.add(section);

            for (int j = 0; j < transactions.size(); j++) {
                if (section.header.equalsIgnoreCase(DateUtils.getDataDayMonth(transactions.get(j).getDate()))) {
                    section.items.add(transactions.get(j));
                }
            }
        }
    }

    private List<String> datesSectionList() {
        List<String> sectionLabels = new ArrayList<>();
        for (int i = 0; i < transactions.size(); i++) {
            sectionLabels.add(DateUtils.getDataDayMonth(transactions.get(i).getDate()));
        }
        return sectionLabels;
    }

    @Override
    public void onBindItemViewHolder(SectioningAdapter.ItemViewHolder viewHolder, int sectionIndex, final int itemIndex, int itemType) {
        final Transaction s = sections.get(sectionIndex).items.get(itemIndex);
        final TransactionItemViewHolder holder = (TransactionItemViewHolder) viewHolder;

        if (s.getType().equals("Income")) {
            holder.textViewAmount.setText(String.format("+%s", NumberUtils.getFormattedAmount(s.getAmount())));
            holder.textViewAmount.setTextColor(ContextCompat.getColor(context, R.color.colorIncome));
        } else {
            holder.textViewAmount.setText("-" + NumberUtils.getFormattedAmount(s.getAmount()));
            holder.textViewAmount.setTextColor(ContextCompat.getColor(context, R.color.colorExpense));
        }
        holder.textViewAmount.setText(NumberUtils.getFormattedAmount(s.getAmount()));
        holder.textViewCategory.setText(s.getCategory());
        holder.textViewType.setText(s.getType());

    }



    @Override
    public void onBindHeaderViewHolder(SectioningAdapter.HeaderViewHolder viewHolder, int sectionIndex, int headerType) {
        Section s = sections.get(sectionIndex);
        HeaderSectionViewHolder header = (HeaderSectionViewHolder) viewHolder;
        header.textViewSection.setText(s.header);
    }

    @Override
    public int getNumberOfSections() {
        return sections.size();
    }

    @Override
    public int getNumberOfItemsInSection(int sectionIndex) {
        return sections.get(sectionIndex).items.size();
    }

    @Override
    public boolean doesSectionHaveHeader(int sectionIndex) {
        return !TextUtils.isEmpty(sections.get(sectionIndex).header);
    }

    private class Section {
        int index;
        String header;
        List<Transaction> items = new ArrayList<>();
    }

    public interface OnTransactionClickListener {
        void onTransactionClick(Transaction transaction);
    }

    /*
     *
     * This ViewHolder Hold the transaction Information on click should open Transaction Details Screen
     * */

    public class TransactionItemViewHolder extends SectioningAdapter.ItemViewHolder implements View.OnClickListener {

        @BindView(R.id.textViewCategory)
        TextView textViewCategory;
        @BindView(R.id.textViewAmount)
        TextView textViewAmount;
        @BindView(R.id.textViewType)
        TextView textViewType;

        public TransactionItemViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            int position = getAdapterPosition();
            final int section = TransactionsAdapter.this.getSectionForAdapterPosition(position);
            final int itemIndex = TransactionsAdapter.this.getPositionOfItemInSection(section, position);
            onTransactionClickListener.onTransactionClick(sections.get(section).items.get(itemIndex));
        }

    }

    public class HeaderSectionViewHolder extends SectioningAdapter.HeaderViewHolder {

        @BindView(R.id.textViewSection)
        TextView textViewSection;

        public HeaderSectionViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }
}

File: app/src/main/java/com/udacity/maluleque/meutako/adapters/TransactionsAdapter.kt
package com.udacity.maluleque.meutako.adapters

import android.content.Context
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.content.ContextCompat
import butterknife.BindView
import butterknife.ButterKnife
import com.udacity.maluleque.meutako.R
import com.udacity.maluleque.meutako.model.Transaction
import com.udacity.maluleque.meutako.utils.DateUtils.getDataDayMonth
import com.udacity.maluleque.meutako.utils.NumberUtils.getFormattedAmount
import org.zakariya.stickyheaders.SectioningAdapter
import java.text.DateFormat
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

class TransactionsAdapter(context: Context, private val transactions: List<Transaction>, onTransactionClickListener: OnTransactionClickListener) : SectioningAdapter() {
    private val context: Context
    private val sections: ArrayList<Section>
    private val onTransactionClickListener: OnTransactionClickListener
    override fun onCreateItemViewHolder(parent: ViewGroup, itemType: Int): ItemViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val v = inflater.inflate(R.layout.transaction_item, parent, false)
        return TransactionItemViewHolder(v)
    }

    override fun onCreateHeaderViewHolder(parent: ViewGroup, headerType: Int): HeaderSectionViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val v = inflater.inflate(R.layout.header, parent, false)
        return HeaderSectionViewHolder(v)
    }

    private fun fillSections() {
        val list: List<String> = ArrayList(HashSet(datesSectionList()))
        Collections.sort(list, object : Comparator<String?> {
            var f: DateFormat = SimpleDateFormat("MMM yyyy", Locale.getDefault())
            override fun compare(o1: String?, o2: String?): Int {
                return try {
                    f.parse(o2).compareTo(f.parse(o1))
                } catch (e: ParseException) {
                    throw IllegalArgumentException(e)
                }
            }
        })
        for (i in list.indices) {
            val section = Section()
            section.index = i
            section.header = list[i]
            sections.add(section)
            for (j in transactions.indices) {
                if (section.header.equals(getDataDayMonth(transactions[j].date), ignoreCase = true)) {
                    section.items.add(transactions[j])
                }
            }
        }
    }

    private fun datesSectionList(): List<String> {
        val sectionLabels: MutableList<String> = ArrayList()
        for (i in transactions.indices) {
            sectionLabels.add(getDataDayMonth(transactions[i].date))
        }
        return sectionLabels
    }

    override fun onBindItemViewHolder(viewHolder: ItemViewHolder, sectionIndex: Int, itemIndex: Int, itemType: Int) {
        val s = sections[sectionIndex].items[itemIndex]
        val holder = viewHolder as TransactionItemViewHolder
        if (s.type == "Income") {
            holder.textViewAmount!!.text = "+${getFormattedAmount(s.amount)}"
            holder.textViewAmount!!.setTextColor(ContextCompat.getColor(context, R.color.colorIncome))
        } else {
            holder.textViewAmount!!.text = "-${getFormattedAmount(s.amount)}"
            holder.textViewAmount!!.setTextColor(ContextCompat.getColor(context, R.color.colorExpense))
        }
        holder.textViewAmount!!.text = getFormattedAmount(s.amount)
        holder.textViewCategory!!.text = s.category
        holder.textViewType!!.text = s.type
    }

    override fun onBindHeaderViewHolder(viewHolder: HeaderViewHolder, sectionIndex: Int, headerType: Int) {
        val s = sections[sectionIndex]
        val header = viewHolder as HeaderSectionViewHolder
        header.textViewSection!!.text = s.header
    }

    override fun getNumberOfSections(): Int {
        return sections.size
    }

    override fun getNumberOfItemsInSection(sectionIndex: Int): Int {
        return sections[sectionIndex].items.size
    }

    override fun doesSectionHaveHeader(sectionIndex: Int): Boolean {
        return !TextUtils.isEmpty(sections[sectionIndex].header)
    }

    private inner class Section {
        var index = 0
        var header: String? = null
        var items: MutableList<Transaction> = ArrayList()
    }

    interface OnTransactionClickListener {
        fun onTransactionClick(transaction: Transaction?)
    }

    /*
     *
     * This ViewHolder Hold the transaction Information on click should open Transaction Details Screen
     * */
    inner class TransactionItemViewHolder(itemView: View) : ItemViewHolder(itemView), View.OnClickListener {
        @BindView(R.id.textViewCategory)
        lateinit var textViewCategory: TextView

        @BindView(R.id.textViewAmount)
        lateinit var textViewAmount: TextView

        @BindView(R.id.textViewType)
        lateinit var textViewType: TextView

        override fun onClick(view: View) {
            val position = adapterPosition
            val section = getSectionForAdapterPosition(position)
            val itemIndex = getPositionOfItemInSection(section, position)
            onTransactionClickListener.onTransactionClick(sections[section].items[itemIndex])
        }

        init {
            ButterKnife.bind(this, itemView)
            itemView.setOnClickListener(this)
        }
    }

    inner class HeaderSectionViewHolder(itemView: View?) : HeaderViewHolder(itemView) {
        @BindView(R.id.textViewSection)
        lateinit var textViewSection: TextView

        init {
            ButterKnife.bind(this, itemView!!)
        }
    }

    init {
        this.sections = ArrayList()
        this.onTransactionClickListener = onTransactionClickListener
        this.context = context
        fillSections()
    }
}